lib/ae_mdw/node.ex

```diff
@@ -69,188 +79,195 @@ defmodule AeMdw.Node do
   end
 
   @spec previous_aex141_signatures :: %{method_hash() => method_signature()}
-  def previous_aex141_signatures do
+  defmemo previous_aex141_signatures() do
     Contract.previous_aex141_signatures()
     |> map_by_function_hash()
   end
 
   @spec aexn_event_hash_types() :: %{Contracts.event_hash() => aexn_event_type()}
-  def aexn_event_hash_types() do
-    map_event_hash_to_type([
-      :allowance,
-      :approval,
-      :approval_for_all,
-      :burn,
-      :mint,
-      :swap,
-      :edition_limit,
-      :edition_limit_decrease,
-      :template_creation,
-      :template_deletion,
-      :template_mint,
-      :template_limit,
-      :template_limit_decrease,
-      :token_limit,
-      :token_limit_decrease,
-      :transfer
-    ])
+  defmemo aexn_event_hash_types() do
+    map_event_hash_to_type(~w(
+      allowance
+      approval
+      approval_for_all
+      burn
+      mint
+      swap
+      edition_limit
+      edition_limit_decrease
+      template_creation
+      template_deletion
+      template_mint
+      template_limit
+      template_limit_decrease
+      token_limit
+      token_limit_decrease
+      transfer
+    )a)
   end
 
   @spec aexn_event_names() :: %{Contracts.event_hash() => AexnContracts.event_name()}
-  def aexn_event_names() do
+  defmemo aexn_event_names() do
     aexn_event_hash_types()
     |> map_event_hash_to_name()
   end
 
   @spec dex_event_hash_types() :: %{Contracts.event_hash() => aexn_event_type()}
-  def dex_event_hash_types() do
-    map_event_hash_to_type([
-      :pair_created,
-      :swap_tokens
-    ])
+  defmemo dex_event_hash_types() do
+    map_event_hash_to_type(~w(
+      pair_created
+      swap_tokens
+    )a)
   end
 
   @spec dex_event_names() :: %{Contracts.event_hash() => AexnContracts.event_name()}
-  def dex_event_names() do
+  defmemo dex_event_names() do
     dex_event_hash_types()
     |> map_event_hash_to_name()
   end
 
-  @spec hdr_fields(:key | :micro) :: [atom()]
-  def hdr_fields(_arg) do
-    ~w(height prev_hash)a
-  end
-
-  @spec height_proto :: [{non_neg_integer(), non_neg_integer()}]
-  def height_proto do
+  @spec height_proto() :: [{non_neg_integer(), non_neg_integer()}]
+  defmemo height_proto() do
     :aec_hard_forks.protocols() |> Enum.into([]) |> Enum.sort(:desc)
   end
 
-  @spec id_field_type(atom()) :: %{atom() => non_neg_integer()} | nil
-  def id_field_type(_field) do
-    %{}
-  end
-
   @spec id_fields :: MapSet.t()
-  def id_fields do
-    MapSet.new()
-  end
-
-  @spec id_prefix(binary()) :: atom()
-  def id_prefix(_arg) do
-  end
-
-  @spec id_prefixes :: MapSet.t()
-  def id_prefixes do
-    MapSet.new()
+  defmemo id_fields() do
+    {_tx_field_types, _tx_fields, tx_ids} = types_fields_ids()
+
+    tx_ids
+    |> Enum.reduce(%{}, fn {type, ids_map}, acc ->
+      for {field, pos} <- ids_map,
+          reduce: acc,
+          do: (acc -> Map.put(acc, [field], Map.put(Map.get(acc, [field], %{}), type, pos)))
+    end)
```

Comment:
I think this can be simplified as:
```suggestion
    for {type, ids_map} <- tx_ids,
        {field, pos} <- ids_map,
        reduce: %{},
        do: (acc -> Map.put(acc, [field], Map.put(Map.get(acc, [field], %{}), type, pos)))
```

I also think that using a normal `do` block would make it easier to read